rfcs/0001-messages.md

```diff
@@ -0,0 +1,414 @@
+#[repr(u8)]
+#[non_exhaustive]
+enum MsgType {
+    SharePackage,
+    SigningCommitments,
+    SigningPackage,
+    SignatureShare,
+    AggregateSignature,
+}
+
+/// The numeric values used to identify the protocol version during serialization.
+struct MsgVersion(u8);
+
+const BASIC_FROST_SERIALIZATION: MsgVersion = MsgVersion(0);
+
+/// The numeric values used to identify each participant during serialization.
+///
+/// In the `frost` module, participant ID `0` should be invalid.
+/// But in serialization, we want participants to be indexed from `0..n`,
+/// where `n` is the number of participants.
+/// This helps us look up their shares and commitments in serialized arrays.
+/// So in serialization, we assign the dealer and aggregator the highest IDs,
+/// and mark those IDs as invalid for signers.
+///
+/// "When performing Shamir secret sharing, a polynomial `f(x)` is used to generate
+/// each party’s share of the secret. The actual secret is `f(0)` and the party with
+/// ID `i` will be given a share with value `f(i)`.
+/// Since a DKG may be implemented in the future, we recommend that the ID `0` be declared invalid."
+/// https://raw.githubusercontent.com/ZcashFoundation/redjubjub/main/zcash-frost-audit-report-20210323.pdf#d
+enum ParticipantId {
+    /// A serialized participant ID for a signer.
+    ///
+    /// Must be less than or equal to `MAX_SIGNER_PARTICIPANT_ID`.
+    Signer(u8),
+    /// The fixed participant ID for the dealer.
+    Dealer,
+    /// The fixed participant ID for the aggregator.
+    Aggregator,
+}
+
+/// The fixed participant ID for the dealer.
+const DEALER_PARTICIPANT_ID: u8 = u8::MAX - 1;
+
+/// The fixed participant ID for the aggregator.
+const AGGREGATOR_PARTICIPANT_ID: u8 = u8::MAX;
+
+/// The maximum `ParticipantId::Signer` in this serialization format.
+///
+/// We reserve two participant IDs for the dealer and aggregator.
+const MAX_SIGNER_PARTICIPANT_ID: u8 = u8::MAX - 2;
+```
+
+### Payloads
+
+Each payload defines a new message:
+
+```rust
+/// The data required to serialize `frost::SharePackage`.
+///
+/// The dealer sends this message to each signer for this round.
+/// With this, the signer should be able to build a `SharePackage` and use
+/// the `sign()` function.
+///
+/// Note: `frost::SharePackage.public` can be calculated from `secret_share`.
+struct messages::SharePackage {
+    /// The public signing key that represents the entire group:
+    /// `frost::SharePackage.group_public`.
+    group_public: VerificationKey<SpendAuth>,
+    /// This participant's secret key share: `frost::SharePackage.share.value`.
+    secret_share: frost::Scalar,
+    /// Commitment for the signer as a single jubjub::AffinePoint.
+    /// A set of commitments to the coefficients (which themselves are scalars)
+    /// for a secret polynomial _f_: `frost::SharePackage.share.commitment`
+    share_commitment: Vec<jubjub::AffinePoint>,
```

Comment:
I think this can be https://github.com/ZcashFoundation/redjubjub/blob/main/src/frost.rs#L91?

```suggestion
    share_commitment: frost::ShareCommitment,
```

Reply:
It is smaller the way we are doing. `ShareCommitment` (https://github.com/ZcashFoundation/redjubjub/blob/main/src/frost.rs#L91) is a vector of `Commitment` (https://github.com/ZcashFoundation/redjubjub/blob/main/src/frost.rs#L76) which is an `jubjub::ExtendedPoint`. We prefer to use a vector of `AffinePoint` instead as it is smaller in size.

Reply:
...
We then convert the `AffinePoint` to/from 32 byte representation https://github.com/ZcashFoundation/redjubjub/blob/74df415b1f95c0d297ebe36c67812f9c6bf9d8f7/rfcs/0001-messages.md#affinepoint

Reply:
So looking at all usages of `generate_shares()` and `verify_share()` currently, I think we should change the `Commitment`s to be newtypes around `jubjub::AffinePoint`, so as to avoid this end-run around collective types like `ShareCommitment`, which exists for this use. The transform into `jubjub::ExtendedPoint` happens once per commitment before compute is done, which is fine / not expensive. 

@chelseakomlo any thoughts

Reply:
I made this changes at https://github.com/ZcashFoundation/redjubjub/pull/85/commits/83794c56bb6d899f765bbbef6f20a70f6a5d55cc